prov/usnic/src/usdf_cq.c

```diff
@@ -650,12 +678,49 @@ usdf_cq_strerror(struct fid_cq *eq, int prov_errno, const void *err_data,
 	return fi_strerror(prov_errno);
 }
 
+/* Handle the associated wait object when closing a CQ.
+ * - Remove the FD from the wait set epoll context
+ * - Decrement the ref count on the wait set
+ * - Remove the CQ from the CQ list attached to the wait set
+ */
+static int usdf_cq_unbind_wait(struct usdf_cq *cq)
+{
+	int ret;
+	struct usdf_wait *wait_priv;
+	struct epoll_event event = {0};
+
+	if (!cq->cq_attr.wait_set) {
+		USDF_DBG_SYS(CQ, "can't unbind from non-existent wait set\n");
+		return -FI_EINVAL;
+	}
+
+	wait_priv = wait_ftou(cq->cq_attr.wait_set);
+
+	ret = epoll_ctl(wait_priv->object.epfd, EPOLL_CTL_DEL,
+			cq->object.fd, &event);
+	if (ret) {
+		USDF_WARN_SYS(CQ,
+				"failed to remove FD from wait set\n");
+		return -errno;
+	}
+
+	fid_list_remove(&wait_priv->list, &wait_priv->lock, &cq->cq_fid.fid);
```

Comment:
I guess my 2 questions distill down to: if there's 2 different atomic actions occurring separately, does that really protect you?  I.e., do *both* need to happen or *neither* need to happen?  Or is there a race condition such that something "bad" could happen between the two protected actions?